### DISCUSSION TITLE
Day 2 morning library exercise bug

### DISCUSSION THREAD

**Discussion by njr0** [google/comprehensive-rust#1233] [General]
As far as I can see, the template code and solution both have a bug.

When `library` is created, it needs to be declared as mutable in order for `add_book` to work:

i.e

    let mut library = Library::new();

rather than

    let library = Library::new();

Both my own solution and the model solution seem to fail without this:

```
cargo run
   Compiling ex19_1_model v0.1.0 (/Users/njr/rust/comp-rust-examples/ex19_1_model)
error[E0596]: cannot borrow `library` as mutable, as it is not declared as mutable
  --> src/main.rs:71:9
   |
71 |     let library = Library::new();
   |         ^^^^^^^ not mutable
...
75 |     library.add_book(Book::new("Lord of the Rings", 1954));
   |     ------------------------------------------------------ cannot borrow as mutable
76 |     library.add_book(Book::new("Alice's Adventures in Wonderland", 1865));
   |     --------------------------------------------------------------------- cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
71 |     let mut library = Library::new();
   |         +++

For more information about this error, try `rustc --explain E0596`.
error: could not compile `ex19_1_model` (bin "ex19_1_model") due to previous error
```

**Comment by mgeisler** [comment]
Hi @njr0, you're spot on that `library` needs to be mutable — that's actually meant to be part of the exerise, so well done for solving it!

Joking aside, we've added a bit more commentary to this exercise over time since it wasn't clear at all what you can change and what you should leave alone. Right now, the hint about this is in the comment saying

> You may also need to update the variable bindings within main.

That might be too subtle... could you make a PR to update the text before the big code listing to explain this better?

**Comment by njr0** [comment]
Thanks for coming back quickly, Martin (@mgeisler).

To be honest, in my view, updating the text will make no difference because I suspect a overwhelming proportion of people (including me) won't read that text.

It seems to me that the "contract" with these exercises is that the "desired behaviour" defines how it's supposed to work. That's what I learned in all the previous exercises. I think it's pedagologically duvbious to say (as you do)

    This shows the desired behavior

and then use *later comments* to attempt to qualify that unqualified statement ("just kidding'!"). But it seems wrong anyway: you're just leaving the learner on quicksand. If I get to change `main`, maybe I'll just change it to print `"Hello, World!"` and be done with it.

Even if you *do* want the exercise to be "Here's the kind of thing you're supposed to make the code do, but feel free to change it however you need so it works", I *still* think your model answer should include the necessary change.

So (with apologies) no, I won't make a PR to update the text, because I don't think the text is the problem.

I'll make a PR to change main to include the `mut` if you like ;-).

**Reply by mgeisler** [reply]
Thanks, you raise excellent point! I certainly don't want to trick people like that :smile:

**Comment by djmitche** [comment]
This raises an interesting issue that I've butted my head against a few times. We want students to get some familiarity with debugging errors from the compiler, since that comprises a greater portion of coding time in Rust than most other languages (made up for with less time typing and less time spent runtime debugging). So the idea in this exercise is to make the suggested implementation, _including_ changing the given code as necessary, but I can certainly see why it feels like the given code can't be changed.

Maybe I'm thinking of exercises as a chance to practice: a time- and complexity-bounded opportunity to work with Rust. If students want to take them in an unexpected direction, or not do them at all, that's fine! But if they're approached from the more academic perspective of homework or exam questions, then questions like "are we allowed to .." or "is it cheating to .." are more sensible.

I'd be interested in ideas as to how we could push students toward considering these to be "practice sessions".

**Reply by mgeisler** [reply]
> Maybe I'm thinking of exercises as a chance to practice: a time- and complexity-bounded opportunity to work with Rust. If students want to take them in an unexpected direction, or not do them at all, that's fine! But if they're approached from the more academic perspective of homework or exam questions, then questions like "are we allowed to .." or "is it cheating to .." are more sensible.

I think there is a big difference between browing the course at home and in a classroom — in a classroom, the instructor can directly tell the students that it's fine to play around, but this is missing in the "offline" version where you study by yourself. So we should do a better job at explaining such things (or avoiding having to explain them).

**Comment by djmitche** [comment]
Oh, and the solution should definitely have the `mut`! Most of the exercise solutions don't have commented-out code in the solution, so this is an outlier in that context.

**Reply by mgeisler** [reply]
Let me fix that real quick :slightly_smiling_face:

**Comment by njr0** [comment]
> This raises an interesting issue that I've butted my head against a few times. We want students to get some familiarity with debugging errors from the compiler, since that comprises a greater portion of coding time in Rust than most other languages (made up for with less time typing and less time spent runtime debugging). So the idea in this exercise is to make the suggested implementation, _including_ changing the given code as necessary, but I can certainly see why it feels like the given code can't be changed.

I completely agree with that, and I think the rust compiler errors are one of the best features of rust.

But I think if you're going to make it so that the main code isn't necessarily right, you need to establish that really early on. And I think anything you want people to fix that way, you can do in the code they (think they're) supposed to edit, not in the "This code defines the interface" (-like) `main` section.

**Reply by mgeisler** [reply]
Hey Nick, what do you think of #1237? It seemed like a good idea to let people pick the right kind of `self` parameter, but on second though, I think it's more instructive to show people the right parameter type and then let them focus on the method bodies.
